### PR TITLE
test: Integration Tests for Sync Watch with Nested Stacks

### PR DESCRIPTION
**Pull request by mildaniel**
#### Why is this change necessary?
Adds integration tests validating `sam sync --watch` for infra and code changes with nested stacks.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write/update unit tests
- [x] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: tests/integration/sync/test_sync_watch.py [hunk 1]
```diff
@@ -67,11 +65,15 @@ def setUp(self):
         self.stacks = []
         self.s3_prefix = uuid.uuid4().hex
         self.test_dir = Path(tempfile.mkdtemp())
+        self.template_before = "" if not self.template_before else self.template_before
```
Comment (mndeveci) [inline tests/integration/sync/test_sync_watch.py:68]: What would be the case when `self.template_before` would be an empty string?

File: tests/integration/sync/test_sync_watch.py [hunk 2]
```diff
@@ -128,78 +127,178 @@ def test_sync_watch_infra(self, runtime):
...
+            with open(destination, "wb") as destination_file:
+                destination_file.write(source_file.read())
+
+
+@parameterized_class([{"runtime": "python"}])
+class TestSyncCodeInfra(TestSyncWatchBase):
+    @classmethod
+    def setUpClass(cls):
+        cls.template_before = f"infra/template-{cls.runtime}-before.yaml"
+        cls.dependency_layer = True
```
Comment (mndeveci) [inline tests/integration/sync/test_sync_watch.py:158]: We could also parameterize this to extend integration test coverage.

**Review by mndeveci — COMMENTED** [review]
Looks good in general left some comments to understand some details and also extending test coverage by parameterizing ADL flag.

File: tests/integration/sync/test_sync_watch.py (hunk 1 above)
Comment (mildaniel) [reply tests/integration/sync/test_sync_watch.py:68]: It should always be defined in the first place. I just initialized it here to appease the IDE.

File: tests/integration/sync/test_sync_watch.py [hunk 3]
```diff
@@ -150,12 +150,13 @@ def update_file(source, destination):
                 destination_file.write(source_file.read())
 
 
-@parameterized_class([{"runtime": "python"}])
+@parameterized_class(
+    [{"runtime": "python", "dependency_layer": True}, {"runtime": "python", "dependency_layer": False}]
```
Comment (mndeveci) [inline tests/integration/sync/test_sync_watch.py:154]: 🥳

**Review by mndeveci — APPROVED** [review]

**Review by qingchm — APPROVED** [review]
LGTM

**Pull request merged**